File: src/53_Maximum_Subarray.py
# 53_Maximum_Subarray
# https://leetcode.cn/problems/maximum-subarray/

from typing import List

# 超时
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        pre_sum = [0] + [sum(nums[:i+1]) for i in range(len(nums))]
        print(pre_sum)
        ans = -100000
        for i in range(len(nums), 0, -1):
            # print(pre_sum[:i])
            ans = max(ans, pre_sum[i] - min(pre_sum[:i]))
        
        return ans

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        ans = -inf
        min_pre_sum = pre_sum = 0
        for x in nums:
            pre_sum += x  # 当前的前缀和
            ans = max(ans, pre_sum - min_pre_sum)  # 减去前缀和的最小值
            min_pre_sum = min(min_pre_sum, pre_sum)  # 维护前缀和的最小值
        return ans

# DP
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if not nums:
            return 0

        # Initialize the dp array
        dp = [0] * len(nums)
        dp[0] = nums[0]

        # Fill the dp array
        for i in range(1, len(nums)):
            dp[i] = max(nums[i], dp[i-1] + nums[i])
        
        # Return the maximum value in dp
        return max(dp)

    
        

        
s = Solution()
ans = s.maxSubArray([5,4,-1,7,8])
print(ans)

File: src/53_Maximum_Subarray.rs
// 53_Maximum_Subarray
// https://leetcode.cn/problems/maximum-subarray/

// 53. 最大子数组和
// 给你一个整数数组 nums ，请你找出一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。
// 子数组 是数组中的一个连续部分。
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        if nums.iter().filter(|&x| *x < 0).count() == nums.len() {
            let mut ns = nums.clone();
            ns.sort_unstable();
            return ns[nums.len() - 1];
        }
        let mut pre_sum = nums[0];
        let mut max = i32::MIN;
        for i in 1..nums.len() {
            pre_sum = pre_sum.max(0);
            max = max.max(pre_sum);
            pre_sum += nums[i];
        }
        max = max.max(pre_sum);

        max
    }
}

use std::cmp;
impl Solution {
    pub fn max_sub_array(nums: Vec<i32>) -> i32 {
        let mut dp = vec![0; nums.len()];
        dp[0] = nums[0];
        let mut ans = nums[0];
        for i in 1..nums.len() {
            if dp[i - 1] <= 0 {
                dp[i] = nums[i];
            } else {
                dp[i] = dp[i - 1] + nums[i];
            }
            ans = cmp::max(ans, dp[i]);
        }
        ans
    }
}
